subxt/src/blocks/blocks_client.rs

```diff
@@ -0,0 +1,151 @@
+#[derive(Derivative)]
+#[derivative(Clone(bound = "Client: Clone"))]
+pub struct BlocksClient<T, Client> {
+    client: Client,
+    _marker: PhantomDataSendSync<T>,
+}
+
+impl<T, Client> BlocksClient<T, Client> {
+    /// Create a new [`BlocksClient`].
+    pub fn new(client: Client) -> Self {
+        Self {
+            client,
+            _marker: PhantomDataSendSync::new(),
+        }
+    }
+}
+
+impl<T, Client> BlocksClient<T, Client>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    /// Subscribe to new best block headers.
+    ///
+    /// # Note
+    ///
+    /// The best block is selected by the consensus algorithm.
+    ///
+    /// These blocks haven't necessarily been finalised yet. Prefer
+    /// [`BlocksClient::subscribe_finalized_headers()`] if that is important.
+    pub fn subscribe_headers(
+        &self,
+    ) -> impl Future<Output = Result<impl Stream<Item = Result<T::Header, Error>>, Error>>
+           + Send
+           + 'static {
+        let client = self.client.clone();
+        async move { client.rpc().subscribe_blocks().await }
+    }
+
+    /// Subscribe to finalized block headers.
+    ///
+    /// While the Substrate RPC method does not guarantee that all finalized block headers are
+    /// provided, this function does.
+    /// ```
+    pub fn subscribe_finalized_headers(
+        &self,
+    ) -> impl Future<Output = Result<impl Stream<Item = Result<T::Header, Error>>, Error>>
+           + Send
+           + 'static {
+        let client = self.client.clone();
+        async move { subscribe_finalized_headers(client).await }
+    }
+}
+
+async fn subscribe_finalized_headers<T, Client>(
+    client: Client,
+) -> Result<impl Stream<Item = Result<T::Header, Error>>, Error>
+where
+    T: Config,
+    Client: OnlineClientT<T>,
+{
+    // Fetch the last finalised block details immediately, so that we'll get
+    // all blocks after this one.
+    let last_finalized_block_hash = client.rpc().finalized_head().await?;
+    let last_finalized_block_num = client
+        .rpc()
+        .header(Some(last_finalized_block_hash))
+        .await?
+        .map(|h| (*h.number()).into());
+
+    let sub = client.rpc().subscribe_finalized_blocks().await?;
+
+    // Adjust the subscription stream to fill in any missing blocks.
+    Ok(subscribe_to_block_headers_filling_in_gaps(
+        client.rpc().clone(),
+        last_finalized_block_num,
+        sub,
+    )
+    .boxed())
+}
+
+/// Note: This is exposed for testing but is not considered stable and may change
+/// without notice in a patch release.
+#[doc(hidden)]
+pub fn subscribe_to_block_headers_filling_in_gaps<T, S, E>(
+    client: Rpc<T>,
+    mut last_block_num: Option<u64>,
+    sub: S,
+) -> impl Stream<Item = Result<T::Header, Error>> + Send
+where
+    T: Config,
+    S: Stream<Item = Result<T::Header, E>> + Send,
+    E: Into<Error> + Send + 'static,
+{
+    sub.flat_map(move |s| {
+        let rpc = client.clone();
```

Comment:
I don't recall off the top of my head why there is a clone here. My guess is removing the move stuff would introduce a lifetime on `client` which wouldn't work out since we want to return that thing. Perhaps we can avoid a little cloning somewhere though?

Reply:
Couldn't remove the clone from here, but removed one clone from above `client.rpc().clone()` when calling this function. The clone was circumvented by passing in a `Client` trait directly and then using the `.rpc()` inside the fn.